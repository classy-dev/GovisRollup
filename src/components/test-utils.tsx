import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeGovisProvider } from "./ThemeGovisProvider";

// 모든 컴포넌트 테스트를 위한 커스텀 렌더 함수
const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => {
  const Wrapper = ({ children }: { children: React.ReactNode }) => {
    return <ThemeGovisProvider>{children}</ThemeGovisProvider>;
  };

  return render(ui, { wrapper: Wrapper, ...options });
};

// re-export testing-library의 나머지 함수들
export * from "@testing-library/react";

// customRender로 오버라이드
export { customRender as render };
