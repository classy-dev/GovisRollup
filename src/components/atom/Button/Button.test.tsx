import { screen, fireEvent } from "@testing-library/react";
import { render } from "@ComponentFarm/test-utils";
import { Button, BtnDelete } from "./Button";
import { FaCheck, FaTimes } from "react-icons/fa";

describe("<Button />", () => {
  /** 1) 텍스트 렌더링 */
  it("children 이 정상적으로 표시된다", () => {
    render(<Button>CONFIRM</Button>);
    expect(screen.getByText("CONFIRM")).toBeInTheDocument();
  });

  /** 2) onClick – 활성 / 비활성 */
  it("클릭 시 onClick 이 호출된다 (enabled)", () => {
    const handle = jest.fn();
    render(<Button onClick={handle}>CLICK</Button>);
    fireEvent.click(screen.getByRole("button"));
    expect(handle).toHaveBeenCalledTimes(1);
  });

  it("disabled 상태에선 onClick 이 호출되지 않는다", () => {
    const handle = jest.fn();
    render(
      <Button onClick={handle} disabled>
        DISABLED
      </Button>
    );
    fireEvent.click(screen.getByRole("button"));
    expect(handle).not.toHaveBeenCalled();
  });

  /** 3) isLoading → Spinner 표시, children 숨김 */
  it("isLoading 이 true면 Spinner만 렌더링된다", () => {
    render(<Button isLoading>LOADING</Button>);
    expect(screen.queryByText("LOADING")).toBeNull();
    expect(screen.getByLabelText("loading")).toBeInTheDocument();
  });

  /** 4) 아이콘 조합 */
  it("Leading, Trailing, IconOnly 렌더링", () => {
    render(
      <Button
        LeadingIcon={<FaCheck data-testid="lead" />}
        TrailingIcon={<FaTimes data-testid="trail" />}
      >
        ICONS
      </Button>
    );
    expect(screen.getByTestId("lead").tagName).toBe("svg");
    expect(screen.getByTestId("trail").tagName).toBe("svg");

    render(<Button IconOnly={<FaCheck data-testid="only" />} />);
    expect(screen.getByTestId("only").tagName).toBe("svg");
  });

  /** 5) 대표 variant+size 스냅샷 (회귀 테스트) */
  it.each`
    variant          | size
    ${"primary"}     | ${"md"}
    ${"gostPrimary"} | ${"sm"}
    ${"transparent"} | ${"lg"}
  `("snapshot – %s / %s", ({ variant, size }) => {
    const { container } = render(
      <Button variant={variant as any} size={size as any}>
        SNAP
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("<BtnDelete />", () => {
  it("onClick 이 잘 호출되며 이벤트 전파를 중단한다", () => {
    const del = jest.fn();
    const parentClick = jest.fn();

    render(
      <div onClick={parentClick}>
        <BtnDelete onClick={del} />
      </div>
    );

    fireEvent.click(screen.getByRole("button"));
    expect(del).toHaveBeenCalledTimes(1);
    expect(parentClick).not.toHaveBeenCalled();
  });
});
