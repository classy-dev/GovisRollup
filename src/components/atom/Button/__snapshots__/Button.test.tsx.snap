// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> snapshot – %s / %s 1`] = `
<button
  class="css-8rp3t8"
  type="button"
>
  <span
    class="txt"
  >
    SNAP
  </span>
</button>
`;

exports[`<Button /> snapshot – %s / %s 2`] = `
<button
  class="css-z5osdt"
  type="button"
>
  <span
    class="txt"
  >
    SNAP
  </span>
</button>
`;

exports[`<Button /> snapshot – %s / %s 3`] = `
<button
  class="css-1qbi09v"
  type="button"
>
  <span
    class="txt"
  >
    SNAP
  </span>
</button>
`;
